utils/state.py

```diff
@@ -0,0 +1,64 @@
+import os
+
+from botocore.errorfactory import ClientError
+
+from utils.aws_api import AWSApi
+
+
+class State(object):
+    """
+    A state object to be used by stateful integrations.
+    A stateful integration is one that has to do each action only once,
+    and there is no source of truth to validate against.
+
+    Good example: email-sender should only send each email once
+    Bad example: openshift-resources' source of truth is the clusters
+
+    :param integration: name of calling integration
+    :param accounts: Graphql AWS accounts query results
+    :param settings: App Interface settings
+
+    :type integration: string
+    :type accounts: list
+    :type settings: dict
+    """
+    def __init__(self, integration, accounts, settings=None):
+        """Initiates S3 client from AWSApi."""
+        self.state_path = f"state/{integration}"
+        self.bucket = os.environ['APP_INTERFACE_STATE_BUCKET']
+        account = os.environ['APP_INTERFACE_STATE_BUCKET_ACCOUNT']
+        aws_api = AWSApi(1, accounts, settings=settings)
+        session = aws_api.get_session(account)
+
+        self.client = session.client('s3')
+
+    def exists(self, key):
+        """checks if a key exists in the state
+
+        Arguments:
+            key {string} -- key to check
+
+        Returns:
+            bool -- True if key exists in state, else False
+        """
+        try:
+            self.client.head_object(
+                Bucket=self.bucket, Key=f"{self.state_path}/{key}")
+            return True
+        except ClientError:
+            return False
+
+    def add(self, key):
+        """adds a key to the state and fails if the key already exists
+
+        Arguments:
+            key {string} -- key to add
+
+        Raises:
+            KeyError: key already exists in the state
+        """
+        if self.exists(key):
+            raise KeyError(
+                f"[state] key {key} already exists in {self.state_path}")
+        self.client.put_object(
+            Bucket=self.bucket, Key=f"{self.state_path}/{key}")
```

Comment:
I like this class. It's clean and direct to the point. I thinking... how about using `__setitem__` and `__getitem__` to implement the functionality here? Anyway, it looks good as is.

Reply:
So, this is what I have in mind:
```python

class Status:
    def __init__(self):
        """
        Status
        """

    def __getitem__(self, item):
        """
        Returns the value of an object:
            client.get_object()['Body'].read()

        If the object body is a JSON serialized content (which I think it's
        a good idea), this returns:
            json.loads(client.get_object()['Body'].read())

        Raises KeyError when the object does not exist.

        Usage:

            foo = status['foo']
        """

    def __setitem__(self, key, value):
        """
        Stores an object. I'd use:
            put_object(..., Body=json.dumps(value))

        Usage:

            status['foo'] = None
        """

    def __delitem__(self, key):
        """
        Deletes an object.

        Raises KeyError when the object does not exist.

        Usage:

            del(status['foo'])
        """

    def __contains__(self, item):
        """
        Checks if an object exists (client.head_object()).

        Usage:

            if 'foo' in status:
                # do something

        Returns True/False.
        """

    def __len__(self):
        """
        Returns the number of objects. Probably the best approach
        is to use boto3.resource.Bucket.objects.all()

        Usage:

            len(status)
        """

    def __iter__(self):
        """
        Creates a generator that yields all the keys. Probably the
        best approach is to use boto3.resource.Bucket.objects.all()

        Usage:

            for key in status:
                print(key)
        """

    def keys(self):
        """
        Returns list(self), which takes advantage of __iter__.

        Usage:

            for key in status.keys():
                print(key)

        """

    def values(self):
        """
        Returns [self[key] for key in self.keys()], which takes advantage
        of __getitem__ and self.keys().

        Usage:

            for value in status.values():
                print(value)
        """

    def items(self):
        """
        For key in self.keys(), this yields (key, self[key]).

        Usage:

            for key, value in status.items():
                print(key, value)
        """
```

Reply:
I totally agree with this suggestion and the proposed solution.
that said, we currently only maintain a state of happened/did not happen. this will be the first feature we do when we have an integration that needs state with content. i've added an issue for it: #419.

I'll keep this class as is, as your first comment says that as well :)

Reply:
Sure, thank you! I didn't mean to implement all the interfaces... just those needed at this point, which are the `add()` and the `__contains__`. Thak you for creating the issue.